components/brave_sync/version_helper.cc

```diff
@@ -0,0 +1,68 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_sync/version_helper.h"
+
+#include <map>
+#include <string>
+#include <vector>
+
+#include "base/strings/string_split.h"
+#include "base/version.h"
+
+namespace {
+
+using MajorToFull = std::map<int, base::Version>;
+const MajorToFull can_self_delete_from{
+    {18, base::Version({1, 18, 2})},  // Nightly which actually can self delete
+    // Below are for test, fill with actual when uplifted
+    // {16, base::Version({1, 16, 100})},  // Beta & Dev
+    // {15, base::Version({1, 15, 100})},  // Release
+
+    // For the future
+    {19, base::Version({1, 19, 0})},
+    {20, base::Version({1, 20, 0})},
+    {20, base::Version({1, 21, 0})},
+};
+
+}  // namespace
+
+namespace brave_sync {
+
+// Remove this function when all channels will be able to remove itself
+bool RemoteDeviceSupportsSelfDelete(const std::string& remote_chrome_string) {
+  // Array of versions per each channel which can process delete device request
+  // in a proper way
+  std::vector<base::StringPiece> segments =
+      base::SplitStringPiece(remote_chrome_string, " ", base::TRIM_WHITESPACE,
+                             base::SPLIT_WANT_NONEMPTY);
+  if (segments.empty()) {
+    // Something wrong is in function, assume remove device cannot delete
+    // itself properly
+    return false;
+  }
+
+  base::Version remote_chrome_version(segments[0]);
+  if (!remote_chrome_version.IsValid() ||
+      remote_chrome_version.components().size() != 4u) {
+    return false;
+  }
+
+  const auto& rcvc = remote_chrome_version.components();
+  base::Version remote_brave_version({rcvc[1], rcvc[2], rcvc[3]});
+
+  MajorToFull::const_iterator it =
+      can_self_delete_from.find(remote_brave_version.components()[1]);
```

Comment:
Why are we enumerating this? Shouldn't we just do `bool operator>=(const Version& v1, const Version& v2)` so we know if it is supported.

Reply:
Because if the PR will be uplifted, then a situation is possible when Nightly will have a version above minimum uplifted Beta.

For example PR is merged into Nightly at `1.18.9`. Then it uplifted into Beta at `1.17.100`. In this case simple check version >= `1.17.100` will give unwanted result for `1.18.5` for the user who didn't updated the browser in time, because `1.8.5` is greater than `1.17.100`. This is the reason.

Reply:
When we bump to 2.x.x, this won't work. We should at least check `remote_brave_version.components()[0])`

Reply:
There is alternative way without checking version and it will benefits us in long term
We can have `brave_device_info_specifics.proto` which contains 
```proto
message BraveSpecificFields {
  optional bool is_self_deleted_supported = 1;
}
```

and then import it in `components/sync/protocol/device_info_specifics.proto`
```proto
message DeviceInfoSpecifics {
  // The cache_guid created to identify a sync client on this device.
  optional string cache_guid = 1;

...

  // Device specific information for Sync invalidations.
  optional InvalidationSpecificFields invalidation_fields = 14;

  optional BraveSpecificFields brave_fields = 1000;
}
```
Choosing a big filed number is to reserving for the future changes chromium might add
If we need to add anything in the future, just add it to `BraveSpecificFields`

Reply:
this approach is much better than the versions check. Thanks, implementing

Reply:
Redone at https://github.com/brave/brave-core/pull/6732/commits/4896efa633008f0ab0d435d878b0528a74b46f98 and https://github.com/brave/brave-core/pull/6732/commits/c991c510d24eac8fef005c7bb4143bb871ecab7a